base/c.jl

```diff
@@ -503,3 +503,146 @@ end
+    firstarg = call.args[2]
+    if firstarg isa Expr && firstarg.head === :parameters
+        length(firstarg.args) > 1 || (kw = firstarg.args[1].args)[1] !== :varargs &&
+            error("@ccall only takes one keyword argument: varargs")
+        hasvarargs = true
+        vararg_type = kw[2]
+        return vararg_type, hasvarargs, call.args[3:end]
+    else
+        return Nothing, hasvarargs, call.args[2:end]
+    end
+end
+
+"""
+       mkarg(arg)
+
+`mkarg` takes an argument and returns a tuple of symbol and type. The
+argument must have a type annotation or an error will be thrown.
+"""
+function mkarg(arg)
+    if arg.head != :(::)
+        error("args in @ccall need type annotations")
+    end
+    return (arg=arg.args[1], type=arg.args[2])
+end
+
+"""
+    mkargs(args, has_varargs::Bool)
+
+takes an iterable of arguments from a function call and returns an
+array of named tuples of (arg=symbol, type=type). If `has_varargs` is
+set to `true`, the value of `type` will be set to `NoType`, to be
+filtered out later.
+"""
+function mkargs(args, has_varargs)
+    in_vararg::Bool = false
+    return map(args) do arg
+        in_vararg && return (arg=arg, type=NoType)
+        if has_varargs && !isa(arg, Expr)
+            in_vararg = true
+            return (arg=arg, type=NoType)
+        end
+        return mkarg(arg)
+    end
+end
+
+"""
+    parsecall(expression)
+
+`parsecall` is an implementation detail of `@ccall
+
+it takes and expression like `:(printf("%d"::Cstring, value::Cuint)::Cvoid)`
+returns: a tuple of `(function_name, return_type, arg_types, args)`
+
+The above input outputs this:
+
+    (:printf, :Cvoid, :((Cstring, Cuint)), ["%d", :value])
+
+Note that the args are in an array, not a quote block and have to be
+appended to the ccall in a separate step.
+"""
+function parsecall(expr)
+    # setup and check for errors
+    expr.head != :(::) &&
+        error("@ccall needs a function signature with a return type")
+    rettype = expr.args[2]
+
+    call = expr.args[1]
+    call.head != :call &&
+        error("@ccall has to be a function call")
+
+    # get the function symbols
+    func = getcfunc(call.args[1])
+    vararg_type, hasvarargs, allargs = getargs(call)
+
+    # separate annotations from names
+    pairs = mkargs(allargs, hasvarargs)
+    args = [a.arg for a in pairs]
+    types = Any[a.type for a in pairs if a.type !== NoType]
+    hasvarargs && push!(types, :($vararg_type...))
+    argtypes = :(())
+    argtypes.args = types
+    return func, rettype, argtypes, args
+end
+
+"""
+    @ccall(call expression)
+
+convert a julia-style function definition to a ccall:
+
+    @ccall printf("%d"::Cstring, 10::Cint)::Cint
+
+same as:
+
+    ccall(:printf, Cint, (Cstring, Cint), "%d", 10)
+
+All arguments must have type annotations and the return type must also
+be annotated.
+
+varargs are supported with the following convention:
+
+    @ccall printf("%d, %d, %d"::Cstring, 1, 2, 3; varargs=Cint)::Cint
```

Comment:
Couldn't we use `1, 2, 3::Cint...`?

Reply:
Yes. I originally had something more like this, but I opted against it because I didn't like using splatting syntax for something that doesn't really splat. Yes, C also uses `...`, but not quite like this.

I'm open to other conventions, but I hesitate to use splatting syntax in this way because it could be easily misinterpreted.

Reply:
@stevengj as per [this comment](https://github.com/JuliaLang/julia/pull/32748#issuecomment-517299960), I have changed the syntax—though not to what you suggested. I'm marking this conversation as resolved, but your further input about the syntax is definitely welcome.